src/crawl/DeepCrawler/DeepCrawler.js

```diff
@@ -0,0 +1,161 @@
+import { SCRAPE_PAGE_PROMPT, SCRAPE_TYPE_PROMPT, ANALYZE_PAGE_PROMPT, GENERATE_SCHEMA_PROMPT } from './prompts.js';
+
+export class DeepCrawler extends BaseCrawler {
+  // TODO(pilvcc): what is `query` and what do we do with it?
+  async *run(url, query, options) {
+    const fetchOptions = options?.fetchOptions || {};
+
+    const userPrompt = options.originalPrompt;
+
+    const scrapeType = await getPromptScrapeType(userPrompt);
+    console.log(`prompt has scrape type: ${scrapeType}`);
+
+    const schema = await generateSchemaFromPrompt(userPrompt);
+    console.log('schema generated', JSON.stringify(schema, null, 2));
+
+    const seenUrls = new Set();
+    const urlStack = [url];
+
+    while (true) {
+      const latestUrl = urlStack[urlStack.length - 1];
+      console.log('processing', latestUrl);
+
+      // we've seen this url, we're just going in circles now
+      if (seenUrls.has(latestUrl)) return null;
+
+      seenUrls.add(latestUrl);
+
+      const { html } = await this.fetcher.first(latestUrl, fetchOptions);
+
+      const pageInfo = await analyzePage(html, userPrompt);
+      console.log(JSON.stringify(pageInfo, null, 2));
+
+      switch (pageInfo.pageType) {
+        case 'list_view':
+          // now that we're at a list view, use fetcher to do pagination & grab all pages
+          const paginatedDocs = await this.fetcher.fetch(latestUrl, fetchOptions);
+
+          // we're scraping the data from the listview itself, so just return the paginated pages themselves
+          if (pageInfo.hasAllFields || !pageInfo.detailViewUrlSelector || !pageInfo.detailViewUrlAttribute) {
+            for (const doc of paginatedDocs) {
+              yield Promise.resolve({ _url: doc.url });
+            }
+            return;
+          }
+
+          // otherwise, we need to grab the detail views
+
+          for (const doc of paginatedDocs) {
+            const $ = cheerio.load(doc.html);
+            const detailUrls = [];
+
+            $(pageInfo.detailViewUrlSelector).each((_, elem) => {
+              detailUrls.push(this.normalizeUrl($(elem).attr(pageInfo.detailViewUrlAttribute), url));
+            });
+
+            for (const it of detailUrls) {
+              yield Promise.resolve({ _url: it });
+            }
+          }
+
+          return;
+
+        case 'detail_view':
+          if (scrapeType === 'fetch_many') {
+            if (pageInfo.listViewUrl) {
+              urlStack.push(this.normalizeUrl(pageInfo.listViewUrl, url));
+              break;
+            }
+            // if it can't find the list view url, just continue to fetch the current page as a single entry
+          }
+          yield Promise.resolve({ _url: latestUrl });
+          return;
+
+        case 'unknown':
+          if (!pageInfo.guessUrl) return null;
+
+          urlStack.push(this.normalizeUrl(pageInfo.guessUrl, url));
+          break;
+      }
+    }
+  }
+
+  normalizeUrl(url, originalUrl) {
```

Comment:
I think there is a better way to do this, I forget. maybe `URL` class lets you combine stuff?